src/annotator/frame-observer.js

```diff
@@ -97,33 +101,51 @@ export class FrameObserver {
 }
 
 /**
- * Check if we can access this iframe's document
+ * Resolves a Promise when the iframe's document is ready (loaded and parsed)
  *
- * @param {HTMLIFrameElement} iframe
- */
-function isAccessible(iframe) {
-  try {
-    return !!iframe.contentDocument;
-  } catch (e) {
-    return false;
-  }
-}
-
-/**
- * Resolves a Promise when the iframe's DOM is ready (loaded and parsed)
- *
- * @param {HTMLIFrameElement} iframe
+ * @param {HTMLIFrameElement} frame
  * @return {Promise<void>}
+ * @throws {Error} if trying to access a document from a cross-origin iframe
  */
-export function onDocumentReady(iframe) {
-  return new Promise(resolve => {
-    const iframeDocument = /** @type {Document} */ (iframe.contentDocument);
-    if (iframeDocument.readyState === 'loading') {
-      iframeDocument.addEventListener('DOMContentLoaded', () => {
-        resolve();
-      });
-    } else {
-      resolve();
+export function onDocumentReady(frame, counter = 0) {
+  return new Promise((resolve, reject) => {
+    // @ts-expect-error
+    const frameDocument = frame.contentWindow.document;
+    const { readyState, location } = frameDocument;
+
+    // Web browsers initially load a blank document before the final document.
+    // This blank document is (1) accessible, (2) has an empty body and head,
+    // and (3) has a 'complete' readyState, on Chrome and Safari, and an
+    // 'uninitialized' readyState on Firefox. If a blank document is detected and
+    // there is a 'src' attribute, it is expected that the blank document will be
+    // replaced by the final document.
+    if (
+      location.href === 'about:blank' &&
+      frame.hasAttribute('src') &&
+      frame.src !== 'about:blank'
+    ) {
+      setTimeout(() => {
+        // If `onDocumentReady` is recursively called more than 100 times
+        // (10ms * 100 = 1s) exit.
+        if (counter >= 99) {
```

Comment:
This might end up being a little short if the system is slow and under heavy load. We can easily adjust it though.